src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using JsonApiDotNetCore.Configuration;
+using JsonApiDotNetCore.MongoDb.Extensions;
+using JsonApiDotNetCore.MongoDb.Queries.Internal.QueryableBuilding;
+using JsonApiDotNetCore.Repositories;
+using JsonApiDotNetCore.Resources;
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+using JsonApiDotNetCore.Queries;
+using JsonApiDotNetCore.Queries.Expressions;
+using JsonApiDotNetCore.Resources.Annotations;
+
+namespace JsonApiDotNetCore.MongoDb
+{
+    public class MongoEntityRepository<TResource, TId>
+        : IResourceRepository<TResource, TId>
+        where TResource : class, IIdentifiable<TId>
+    {
+        private readonly IMongoDatabase _db;
+        private readonly ITargetedFields _targetedFields;
+        private readonly IResourceGraph _resourceGraph;
+        private readonly IResourceFactory _resourceFactory;
+        private readonly IEnumerable<IQueryConstraintProvider> _constraintProviders;
+
+        public MongoEntityRepository(
+            IMongoDatabase db,
+            ITargetedFields targetedFields,
+            IResourceGraph resourceGraph,
+            IResourceFactory resourceFactory,
+            IEnumerable<IQueryConstraintProvider> constraintProviders)
+        {
+            _db = db;
+            _targetedFields = targetedFields;
+            _resourceGraph = resourceGraph;
+            _resourceFactory = resourceFactory;
+            _constraintProviders = constraintProviders;
+        }
+
+        private IMongoCollection<TResource> Collection => _db.GetCollection<TResource>(typeof(TResource).Name);
+        private IMongoQueryable<TResource> Entities => Collection.AsQueryable();
+
+        public virtual Task<int> CountAsync(FilterExpression topFilter)
+        {
+            var resourceContext = _resourceGraph.GetResourceContext<TResource>();
+            var layer = new QueryLayer(resourceContext)
+            {
+                Filter = topFilter
+            };
+
+            var query = ApplyQueryLayer(layer);
+            return query.CountAsync();
+        }
+
+        public virtual Task CreateAsync(TResource resource) =>
+            Collection.InsertOneAsync(resource);
```

Comment:
Once you've updated to the latest bits of JADNC, be sure to pass along `CancellationToken`s here.